pkg/providers/vsphere/vmprovider_vm_utils.go

```diff
@@ -64,52 +64,60 @@ func GetVirtualMachineImageSpecAndStatus(
 	vmCtx context.VirtualMachineContext,
 	k8sClient ctrlclient.Client) (ctrlclient.Object, *vmopv1.VirtualMachineImageSpec, *vmopv1.VirtualMachineImageStatus, error) {
 
-	var obj conditions.Getter
-	var spec *vmopv1.VirtualMachineImageSpec
-	var status *vmopv1.VirtualMachineImageStatus
-
-	key := ctrlclient.ObjectKey{Name: vmCtx.VM.Spec.ImageName, Namespace: vmCtx.VM.Namespace}
-	vmImage := &vmopv1.VirtualMachineImage{}
-	if err := k8sClient.Get(vmCtx, key, vmImage); err != nil {
-		clusterVMImage := &vmopv1.ClusterVirtualMachineImage{}
-
-		if apierrors.IsNotFound(err) {
-			key.Namespace = ""
-			err = k8sClient.Get(vmCtx, key, clusterVMImage)
+	if vmCtx.VM.Spec.Image == nil {
+		reason := "NotSet"
+		conditions.MarkFalse(vmCtx.VM, vmopv1.VirtualMachineConditionImageReady, reason, "")
+		return nil, nil, nil, fmt.Errorf(reason)
+	}
+
+	var (
+		obj    conditions.Getter
+		objErr error
+		spec   *vmopv1.VirtualMachineImageSpec
+		status *vmopv1.VirtualMachineImageStatus
+		key    = ctrlclient.ObjectKey{
+			Name:      vmCtx.VM.Spec.Image.Name,
+			Namespace: vmCtx.VM.Namespace,
 		}
+	)
 
-		if err != nil {
-			// Don't use the k8s error as-is as we don't know to prefer the NS or cluster scoped error message.
-			// This is the same error/message that the prior code used.
-			reason, msg := "NotFound", fmt.Sprintf("Failed to get the VM's image: %s", key.Name)
-			conditions.MarkFalse(vmCtx.VM, vmopv1.VirtualMachineConditionImageReady, reason, msg)
-			return nil, nil, nil, fmt.Errorf("%s: %w", msg, err)
+	switch vmCtx.VM.Spec.Image.Kind {
+	case "VirtualMachineImage":
+		var img vmopv1.VirtualMachineImage
+		if objErr = k8sClient.Get(vmCtx, key, &img); objErr == nil {
+			obj, spec, status = &img, &img.Spec, &img.Status
 		}
 
-		// Ensure the GVK for the object is synced back into the object since
-		// the object's APIVersion and Kind fields may be used later.
-		if err := kube.SyncGVKToObject(clusterVMImage, k8sClient.Scheme()); err != nil {
-			return nil, nil, nil, err
+	case "ClusterVirtualMachineImage":
+		key.Namespace = ""
+		var img vmopv1.ClusterVirtualMachineImage
+		if objErr = k8sClient.Get(vmCtx, key, &img); objErr == nil {
+			obj, spec, status = &img, &img.Spec, &img.Status
 		}
```

Comment:
Do we want a default case here to at least set `objErr` to something non-nil?

Reply:
Handled in the resend.